avalon/nuke/lib.py

```diff
@@ -0,0 +1,240 @@
+import os
+import contextlib
+import nuke
+import re
+import logging
+import toml
+
+log = logging.getLogger(__name__)
+
+
+@contextlib.contextmanager
+def maintained_selection():
+    nodes = nuke.allNodes()
+    previous_selection = [n.name()
+                          for n in nodes
+                          if n['selected'].value() is True]
+    try:
+        yield
+    finally:
+        # unselect all selection in case there is some
+        [n['selected'].setValue(False) for n in nodes]
+        # and select all previously selected nodes
+        if previous_selection:
+            [n['selected'].setValue(True)
+             for n in nodes
+             if n.name() in previous_selection]
+
+
+def reset_selection():
+    nodes = nuke.allNodes()
+    [n['selected'].setValue(False) for n in nodes]
+
+
+def select_nodes(nodes):
+    assert isinstance(nodes, (list, tuple)), "nodes has to be list or tuple"
+    [n['selected'].setValue(True) for n in nodes]
+
+
+def add_avalon_tab_knob(node):
+    """Adding a tab and a knob into a node
+    """
+    try:
+        avalon_knob = node['avalon'].value()
+    except Exception:
+        tab = nuke.Tab_Knob("Avalon")
+        uk = nuke.Text_Knob('avalon', 'avalon data')
+        node.addKnob(tab)
+        node.addKnob(uk)
+        log.info("created new user knob avalon")
+
+
+def imprint(node, data):
+    """Adding `Avalon data` into a node's Avalon Tab/Avalon knob
+    also including publish knob
+
+    Arguments:
+        node (list or obj): A nuke's node object either in list or individual
+        data (dict): Any data which needst to be imprinted
+    """
+    if not node:
+        return
+    if isinstance(node, list):
+        node = node[0]
+
+    add_avalon_tab_knob(node)
+    add_publish_knob(node)
+    node['avalon'].setValue(toml.dumps(data))
+
+    return node
+
+
+def ls_img_sequence(dirPath, one=None):
```

Comment:
Just wondering, could this be simplified/resolved by using `clique` that is vendorized with avalon core?

Also, if there's a "list image sequence" method and it's non-trivial code (like this) I believe it is somewhat universally needed - could be a good one for `avalon.lib` as opposed to `avalon.nuke.lib`? Or is this specifically targeted at Nuke?

Reply:
Actually I can't figure out why this is even there. It does exactly the same as clique and we shouldn't need it at all.

Reply:
Cool, let's remove and reduce unnecessary code. :)